### PR TITLE
[RPW]Strip Subshaders that are not from the current pipeline

### PR DESCRIPTION
**Pull request by alex-vazquez**
### Purpose of this PR
On each shader stripper, when we know that the Asset is from the correct pipeline, we are able to check the `RenderPipeline` tag, and make sure that it maps with the one defined on the pipeline.

If the tag is not null, but it is different from the current pipeline, it will strip all it's variants.

---
### Testing status
- Checking a cross pipeline project, that when building for HDRP, all URP shaders are stripped
- Checking with a shadergraph targetting HDRP and URP, when building for HDRP, URP Shaders are stripped.
- Adding unit testing
image

---
### Comments to reviewers
This PR was originaly created on: [trunk](https://github.cds.internal.unity3d.com/unity/unity/pull/5947), but I've decided to create it on the package side, as we have more control over the current asset, and we have logging + export information.

### CODE REVIEW

**Review by aleks01010101 — APPROVED** [review]
LGTM

File: com.unity.render-pipelines.core/CHANGELOG.md [hunk 1]
```diff
@@ -11,6 +11,7 @@ and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.
 - Linear version of function that sets FSR RCAS shader constants
 - Add probe volume influence weight parameter
 - Added support for multiple Baking States to Prove Volumes.
+- Method to obtain the `RenderPipeline` tag from a shader given a `ShaderSnippetData`.
```
Comment (RSlysz) [inline com.unity.render-pipelines.core/CHANGELOG.md:14]:
```suggestion
- Add method to obtain the `RenderPipeline` tag from a shader given a `ShaderSnippetData`.
```

**Review by RSlysz — CHANGES_REQUESTED** [review]
The new tag is logged in HDRP but not in Universal. We should have parity if possible to debug more easily.

Also, there is the URP FallbackError shader that got the tag in this PR. What about other shaders? Is it the only one missing a Tag?

File: com.unity.render-pipelines.core/Editor/BuildProcessors/ShaderExtensions.cs [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+        private static readonly ShaderTagId s_RenderPipelineShaderTagId = new ShaderTagId("RenderPipeline");
+
+        /// <summary>
+        /// Tries to find the "RenderPipeline" on the shader by the given <see cref="ShaderSnippetData"/>
+        /// </summary>
+        /// <param name="shader"><see cref="Shader"/> The shader to look for the tag</param>
+        /// <param name="snippetData"><see cref="ShaderSnippetData"/></param>
+        /// <param name="renderPipelineTag"><see cref="string"/> containing the value of the tag "RenderPipeline"</param>
+        /// <returns>true if the tag is found and has a value</returns>
+        public static bool TryGetRenderPipelineTag([NotNull] this Shader shader, ShaderSnippetData snippetData,  out string renderPipelineTag)
```
Comment (RSlysz) [inline com.unity.render-pipelines.core/Editor/BuildProcessors/ShaderExtensions.cs:22]:
```suggestion
        public static bool TryGetRenderPipelineTag([DisallowNull] this Shader shader, ShaderSnippetData snippetData,  out string renderPipelineTag)
```

**Review by RSlysz — APPROVED** [review]

**Review by RSlysz — COMMENTED** [review]
It is possible to have an automated test on this? Regression would be hard to discover

File: com.unity.render-pipelines.high-definition/Editor/BuildProcessors/HDRPPreprocessShaders.cs [hunk 3]
```diff
@@ -483,79 +500,97 @@ public void Dispose()
...
-                        {
-                            removeInput = false;
-                            break;
-                        }
-                    }
+            double stripTimeMs;
+            using (new ExportShaderStrip("Temp/shader-strip.json", shader, snippet, inputData, this))
+            {
+                Stopwatch shaderStripingWatch = new Stopwatch();
+                shaderStripingWatch.Start();
```
Comment (RSlysz) [inline com.unity.render-pipelines.high-definition/Editor/BuildProcessors/HDRPPreprocessShaders.cs:572]: [Optional] If this start to be used elsewhere too, think about making a MesuringScope

File: com.unity.render-pipelines.high-definition/Tests/Editor/BuildProcessors/HDRPPreprocessShaderTests.cs [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+            foreach (var preprocessor in TypeCache.GetTypesDerivedFrom<IPreprocessBuildWithReport>())
+            {
+                try
+                {
+                    var buildReport = Activator.CreateInstance(preprocessor) as IPreprocessBuildWithReport;
+                    buildReport.OnPreprocessBuild(default);
+                }
+                catch
+                {
+                }
```
Comment (RSlysz) [inline com.unity.render-pipelines.high-definition/Tests/Editor/BuildProcessors/HDRPPreprocessShaderTests.cs:25]: Why a Try-Catch discarding all error? If there is an error we will have wrong initialization without knowing it

File: com.unity.render-pipelines.high-definition/Tests/Editor/BuildProcessors/HDRPPreprocessShaderTests.cs [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+using NUnit.Framework;
+using System.Collections.Generic;
+using UnityEngine;
+
+namespace UnityEditor.Rendering.HighDefinition.Tests
+{
+    public class HDRPPreprocessShadersTests
+    {
+        [SetUp]
+        public void InitializeBuildPreprocessors()
```
Comment (RSlysz) [inline com.unity.render-pipelines.high-definition/Tests/Editor/BuildProcessors/HDRPPreprocessShaderTests.cs:13]: I am not sure to understand what is set up here. Can you clarify (by comments if needed)

File: com.unity.render-pipelines.high-definition/Tests/Editor/BuildProcessors/HDRPPreprocessShaderTests.cs [hunk 6]
```diff
@@ -0,0 +1,42 @@
+using System;
+using UnityEditor;
+using UnityEditor.Build;
+using NUnit.Framework;
+using System.Collections.Generic;
+using UnityEngine;
+
+namespace UnityEditor.Rendering.HighDefinition.Tests
+{
+    public class HDRPPreprocessShadersTests
```
Comment (RSlysz) [inline com.unity.render-pipelines.high-definition/Tests/Editor/BuildProcessors/HDRPPreprocessShaderTests.cs:10]: There is a stripping case (StripHDRPShader_WhenHDRPisNotCurrentPipeline).
For completeness, a non stripping case would be nice: DontStripHDRPShader_WhenHDRPisCurrentPipeline

File: com.unity.render-pipelines.universal/Editor/ShaderPreprocessor.cs [hunk 7]
```diff
@@ -664,63 +664,93 @@ void LogShaderVariants(Shader shader, ShaderSnippetData snippetData, int prevVar
...
-                return;
+            if (urpAsset == null)
+                return false;
 
-            m_stripTimer.Start();
+            return true;
+        }
 
-            InitializeLocalShaderKeywords(shader);
+        private bool IsShaderProcessed(Shader shader, ShaderSnippetData snippet)
```
Comment (RSlysz) [inline com.unity.render-pipelines.universal/Editor/ShaderPreprocessor.cs:687]:
```suggestion
        private bool IsURPCompatibleShader(Shader shader, ShaderSnippetData snippet)
```

File: com.unity.render-pipelines.universal/Editor/ShaderPreprocessor.cs [hunk 8]
```diff
@@ -664,63 +664,93 @@ void LogShaderVariants(Shader shader, ShaderSnippetData snippetData, int prevVar
...
 
-                // Remove at swap back
-                if (removeInput)
-                    compilerDataList[i] = compilerDataList[--inputShaderVariantCount];
+            int prevVariantCount = compilerDataList.Count;
+            var outputVariantCount = prevVariantCount;
+
+            if (!active)
+            {
+                if (!IsShaderProcessed(shader, snippetData))
```
Comment (RSlysz) [inline com.unity.render-pipelines.universal/Editor/ShaderPreprocessor.cs:731]:
```suggestion
                if (!IsURPCompatibleShader(shader, snippetData))
```

File: com.unity.render-pipelines.universal/Tests/Editor/BuildProcessors/PreprocessShaderTests.cs [hunk 9]
```diff
@@ -0,0 +1,42 @@
+using System;
+using UnityEditor;
+using UnityEditor.Build;
+using NUnit.Framework;
+using System.Collections.Generic;
+using UnityEngine;
+
+namespace UnityEditor.Rendering.Universal.Tests
+{
+    public class PreprocessShadersTests
```
Comment (RSlysz) [inline com.unity.render-pipelines.universal/Tests/Editor/BuildProcessors/PreprocessShaderTests.cs:10]: sames concerns than the HDRP tests

File: com.unity.render-pipelines.high-definition/Tests/Editor/BuildProcessors/HDRPPreprocessShaderTests.cs [hunk 10]
```diff
@@ -0,0 +1,42 @@
...
+                    buildReport.OnPreprocessBuild(default);
+                }
+                catch
+                {
+                }
+            }
+        }
+
+        [Test]
+        public void StripHDRPShader_WhenHDRPisNotCurrentPipeline()
```
Comment (RSlysz) [inline com.unity.render-pipelines.high-definition/Tests/Editor/BuildProcessors/HDRPPreprocessShaderTests.cs:30]: RenderPipeline is a project global set up. You perhaps should check that you don't use HDRP

File: com.unity.render-pipelines.high-definition/Editor/BuildProcessors/HDRPPreprocessShaders.cs [hunk 11]
```diff
@@ -483,79 +500,97 @@ public void Dispose()
...
-                return;
+                return false;
 
             if (HDRenderPipelineGlobalSettings.Ensure(canCreateNewAsset: false) == null)
-                return;
+                return false;
 
-            Stopwatch shaderStripingWatch = new Stopwatch();
-            shaderStripingWatch.Start();
+            // TODO: Grab correct configuration/quality asset.
```
Comment (RSlysz) [inline com.unity.render-pipelines.high-definition/Editor/BuildProcessors/HDRPPreprocessShaders.cs:514]: don't forget this todo

**Review by RSlysz — CHANGES_REQUESTED** [review]
Minor potential improvement.
The test part should be improved.

**Comment by alex-vazquez** [comment]
Moved to trunk

**Pull request closed without merging**